plenum/test/node_catchup/test_process_catchup_replies.py

```diff
@@ -0,0 +1,156 @@
+        txns = {}
+        for seq_no, txn in ledger.getAllTxn(start, end):
+            txns[str(seq_no)] = ledger_manager.owner.update_txn_with_extra_data(txn)
+        replies.append(CatchupRep(ledger_id,
+                                  SortedDict(txns),
+                                  cons_proof))
+    return ledger_manager._buildConsistencyProof(ledger_id,
+                                                 ledger.seqNo - txn_count,
+                                                 ledger.seqNo), replies
+
+
+def check_reply_not_applied(old_ledger_size, ledger, ledger_info, reply):
+    assert ledger.size == old_ledger_size
+    assert ledger.seqNo == old_ledger_size
+    received_replies = {str(seq_no) for seq_no, _ in ledger_info.receivedCatchUpReplies}
+    assert set(getattr(reply, f.TXNS.nm).keys()).issubset(received_replies)
+
+
+def check_replies_applied(old_ledger_size, ledger, ledger_info, replies):
+    new_txn_count = sum([len(getattr(reply, f.TXNS.nm).keys())
+                         for reply in replies])
+    assert ledger.size == old_ledger_size + new_txn_count
+    assert ledger.seqNo == old_ledger_size + new_txn_count
+    received_replies = {str(seq_no) for seq_no, _ in ledger_info.receivedCatchUpReplies}
+    assert all(not set(getattr(reply, f.TXNS.nm).keys()).issubset(received_replies)
+               for reply in replies)
+
+
+def test_process_catchup_replies(txnPoolNodeSet, looper, sdk_wallet_client):
+    '''
+    Test correct work of method processCatchupRep and that sending replies
+    in reverse order will call a few iterations of cycle in _processCatchupRep
+    '''
+    # count of transactions and replies (one txn in one reply)
+    # and iterations in _processCatchupRep
+    txns_count = 3
+    ledger_manager = txnPoolNodeSet[0].ledgerManager
+    ledger_info = ledger_manager.getLedgerInfoByType(ledger_id)
+    ledger = ledger_manager.ledgerRegistry[ledger_id].ledger
+    old_ledger_size = ledger.size
+    num_txns_in_reply = 3
+    reply_count = 6
+
+    cons_proof, catchup_reps = _add_txns_to_ledger(txnPoolNodeSet[1],
+                                                   looper,
+                                                   sdk_wallet_client,
+                                                   num_txns_in_reply,
+                                                   reply_count)
+    ledger_info.catchUpTill = cons_proof
+    ledger_info.state = LedgerState.syncing
+
+    # send replies in next order: 2, 3, 1
+    # and check that after sending reply 1, replies 2 and 3 will be applied.
+    reply2 = catchup_reps[1]
+    ledger_manager.processCatchupRep(reply2, sdk_wallet_client[1])
+    check_reply_not_applied(old_ledger_size, ledger, ledger_info, reply2)
+
+    reply3 = catchup_reps[2]
+    ledger_manager.processCatchupRep(reply3, sdk_wallet_client[1])
+    check_reply_not_applied(old_ledger_size, ledger, ledger_info, reply2)
```

Comment:
It makes sense to verify that neither `reply2` nor `reply3` is applied.

Reply:
Done